### PR TITLE
ParsedName: add method to generate es name query

### PR DESCRIPTION
**Pull request by MJedr**
* move the functionality from inspire-query-parsed so that it can be reused in other modules

### CODE REVIEW

File: inspire_utils/name.py [hunk 1]
```diff
@@ -246,6 +248,104 @@ def from_parts(
         name.title = title
         return ParsedName(name)
 
+    def generate_es_query(self, author_name, keyword="authors"):
```
Comment (michamos) [inline inspire_utils/name.py:251]: Why is this taking an author name? Don't you already know the name because you've passed it to `ParsedName`? I would also rename `keyword` to `field`, and document everything.

File: inspire_utils/name.py [hunk 2]
```diff
@@ -246,6 +248,104 @@ def from_parts(
...
+                name_query.append(
+                    _match_query_with_names_initials_analyzer_with_and_operator(
+                        '{}.first_name.initials'.format(keyword), name)
+                )
+            else:
+                name_query.extend([
+                    _match_phrase_prefix_query('{}.first_name'.format(keyword), name),
+                    _match_query_with_names_initials_analyzer_with_and_operator('{}.first_name'.format(keyword), name)
+                ])
+            if '.' not in self.first and ',' not in author_name:
```
Comment (michamos) [inline inspire_utils/name.py:332]: Ok, so this is what you need the `author_name` for. This is used in the query parser to determine because there's an ambiguity: is `a John Smith` a search for `first="John", second="Smith"` or `second="John Smith"`? The proper way to have this information is to do this check in the `__init__`, and set a flag such as `maybe_only_last_name` that you can then test here.

File: tests/test_name.py [hunk 3]
```diff
@@ -467,3 +468,779 @@ def test_first_name_initials_without_whitespace_is_initialized_correctly():
 
 def test_last_name_recognized_correctly_regression_test():
     assert u"De Sousa Vieira" == ParsedName.loads('De Sousa Vieira, M.C.').last
+
+
+def test_elastic_search_visitor_author_lastname_firstname_with_commas_and_initials():
```
Comment (michamos) [inline tests/test_name.py:473]:
```suggestion
def test_generate_es_query_lastname_firstname_with_commas_and_initials():
```
and same for all other tests.

File: inspire_utils/name.py [hunk 4]
```diff
@@ -81,9 +83,13 @@ def __init__(self, name, constants=None):
         if isinstance(name, HumanName):
             self._parsed_name = name
         else:
+            self._parsed_name = HumanName(name, constants=constants)
             self._parsed_name = HumanName(name, constants=constants)
             self._parsed_name.capitalize()
 
+        if len(self.first_list) == 1 and '.' not in self.first and ',' not in name:
```
Comment (michamos) [inline inspire_utils/name.py:90]: why the `len` check now? I don't remember it being there previously.

File: inspire_utils/name.py [hunk 5]
```diff
@@ -246,6 +252,104 @@ def from_parts(
...
+                    [
+                        _match_phrase_prefix_query(
+                            u"{}.first_name".format(keyword), name
+                        ),
+                        _match_query_with_names_initials_analyzer_with_and_operator(
+                            u"{}.first_name".format(keyword), name
+                        ),
+                    ]
+                )
+            if hasattr(self, "maybe_only_last_name"):
```
Comment (michamos) [inline inspire_utils/name.py:332]: that's dirty. It should always be there, but be `False` in the other case, so you can simply write
```suggestion
            if self.maybe_only_last_name:
```

File: inspire_utils/query.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+    queries = [q for q in queries if q]
+
+    if len(queries) == 1 and not preserve_bool_semantics_if_one_clause:
+        return queries[0]
+
+    return {"bool": {"must" if use_must_clause else "should": queries}}
+
+
+def ordered(obj):
```
Comment (michamos) [inline inspire_utils/query.py:50]: If you need this only for the tests, you should move it there.

File: inspire_utils/name.py (hunk 4 above)
Comment (MJedr) [reply inspire_utils/name.py:90]: The thing is that `self.first` is only one first name, so names like `James K. Smith` also were marked as True what's wrong

File: inspire_utils/query.py (hunk 6 above)
Comment (MJedr) [reply inspire_utils/query.py:50]: Yes, but it's also used by query parser in tests, I didn't want to duplicate the same function

File: inspire_utils/name.py (hunk 4 above)
Comment (michamos) [reply inspire_utils/name.py:90]: how is that possible, if it contains a `.`? does `self.first` exclude middle names?

Comment (MJedr) [reply inspire_utils/name.py:90]: Yes, it's only first name (one string, not a list)

Comment (michamos) [reply inspire_utils/name.py:90]: but it's not correct: `John David Smith` might be a last name too.

Comment (MJedr) [reply inspire_utils/name.py:90]: So it is wrong now in query parser. I'll change it to create `full_name` query for those cases too.

**Review by michamos — APPROVED** [review]
